### PR TITLE
chapter_11:Account Activations

### PR DESCRIPTION
**Pull request by ngyihkai**
Ready

### CODE REVIEW

**Comment by ngyihkai** [comment]
Ready

File: app/assets/stylesheets/account_activations.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AccountActivations controller here.
+// They will automatically be included in application.css.
```
Comment (ngocvt-0484) [inline app/assets/stylesheets/account_activations.scss]: remove this file because you not use

File: test/mailers/user_mailer_test.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+class UserMailerTest < ActionMailer::TestCase
+  test "account_activation" do
+    mail = UserMailer.account_activation
+    assert_equal "Account activation", mail.subject
+    assert_equal ["to@example.org"], mail.to
+    assert_equal ["from@example.com"], mail.from
+    assert_match "Hi", mail.body.encoded
+  end
+
+  test "password_reset" do
```
Comment (ngocvt-0484) [inline test/mailers/user_mailer_test.rb]: dont need do test. please delete all file on test folder

**Comment by ngyihkai** [comment]
Ready

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -5,13 +5,15 @@ def new
...
     @user = User.find_by(email: params[:session][:email].downcase)
     if @user&.authenticate(params[:session][:password])
-      logged_in @user
-      params[:session][:remember_me] == "1" ? remember(@user) : forget(@user)
-      redirect_back_or @user
-    else
-      flash.now[:danger] = t "invalid_email_password_combination"
-      render :new
-    end
+      if @user.activated
```
Comment (huongnt-2545) [inline app/controllers/sessions_controller.rb:8]: indent from line 8 -> 16

File: app/helpers/account_activations_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
```
Comment (huongnt-2545) [inline app/helpers/account_activations_helper.rb]: delete this file, it's no need

File: app/models/user.rb [hunk 5]
```diff
@@ -21,12 +22,31 @@ def forget
     update_attribute :remember_digest, nil
   end
 
+  def authenticated? attribute, token
+  digest = send "#{attribute}_digest"
```
Comment (huongnt-2545) [inline app/models/user.rb]: indent from line 26 -> 28

File: app/models/user.rb [hunk 6]
```diff
@@ -21,12 +22,31 @@ def forget
...
+  return false if digest.nil?
+  BCrypt::Password.new(digest).is_password? token
+  end
+
   private
```
Comment (huongnt-2545) [inline app/models/user.rb]: please move all methods depend on private scope place at the end of class

Comment (huongnt-2545) [inline app/models/user.rb]: please move all methods depend on private scope place at the end of class

File: config/environments/development.rb [hunk 7]
```diff
@@ -35,6 +35,9 @@
 
   # Don't care if the mailer can't send.
   config.action_mailer.raise_delivery_errors = false
+  config.action_mailer.delivery_method = :test
+  host = 'example.com' # Don't use this literally; use your local dev host instead
```
Comment (huongnt-2545) [inline config/environments/development.rb]:
```suggestion
  host = 'localhost:3000'
```

File: config/environments/development.rb [hunk 8]
```diff
@@ -35,6 +35,9 @@
...
+  config.action_mailer.default_url_options = { host: host, protocol: 'https' }
```
Comment (huongnt-2545) [inline config/environments/development.rb]:
```suggestion
  config.action_mailer.default_url_options = { host: host, protocol: 'http' }
```

File: config/environments/production.rb [hunk 9]
```diff
@@ -117,4 +117,18 @@
   # config.active_record.database_selector = { delay: 2.seconds }
   # config.active_record.database_resolver = ActiveRecord::Middleware::DatabaseSelector::Resolver
   # config.active_record.database_resolver_context = ActiveRecord::Middleware::DatabaseSelector::Resolver::Session
+
+  config.action_mailer.raise_delivery_errors = true
+  config.action_mailer.delivery_method = :smtp
+  host = '<your heroku app>.herokuapp.com'
```
Comment (huongnt-2545) [inline config/environments/production.rb]: using environment variable to save host value
it's the same below:
```
 user_name: ENV['SENDGRID_USERNAME'],
 password: ENV['SENDGRID_PASSWORD']
```

File: test/controllers/account_activations_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
```
Comment (huongnt-2545) [inline test/controllers/account_activations_controller_test.rb]: delete this file, it's no need

File: config/environments/production.rb (hunk 9 above)
Comment (ngyihkai) [reply config/environments/production.rb]: Screenshot from 2021-09-06 17-32-48
is it correct?

**Comment by ngyihkai** [comment]
Ready

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -5,13 +5,15 @@ def new
...
-    end
+        if @user.activated
+          log_in @user
+          params[:session][:remember_me] == "1" ? remember(@user) : forget(@user)
+          redirect_back_or @user
+        else
+          message = t(:error_acc_unacti)
+          flash[:warning] = t(:message)
+          redirect_to root_url
+        end
```
Comment (ghost) [inline app/controllers/sessions_controller.rb]: Action create should be like this pic

File: app/models/user.rb [hunk 12]
```diff
@@ -41,4 +44,21 @@ def digest string
...
+  def downcase_email
+    self.email.downcase!
+  end
+
+  def create_activation_digest
+    self.activation_token = User.new_token
+    self.activation_digest = User.digest(activation_token)
+  end
+
+  def activate
```
Comment (ghost) [inline app/models/user.rb]: can you active an account ?

**Review by ghost — COMMENTED** [review]
cmt

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -45,8 +46,6 @@ def destroy
     redirect_to users_url
   end
 
-  private
```
Comment (huongnt-2545) [inline app/controllers/users_controller.rb]: keep this

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -59,6 +58,8 @@ def load_user
     redirect_to root_path
   end
 
+  private
```
Comment (huongnt-2545) [inline app/controllers/users_controller.rb]: remove this line

**Comment by ngyihkai** [comment]
Ready

File: app/controllers/users_controller.rb (hunk 14 above)
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb]: keep private line 48
it should not here

**Comment by ngyihkai** [comment]
Ready

**Review by huongnt-2545 — APPROVED** [review]
merged
you can merge this pull

**Pull request merged**